### DISCUSSION TITLE
Can I change the property used for validation?

### DISCUSSION THREAD

**Discussion by mikemand** [livewire/livewire#4459] [1. Help] [answered]
For larger datasets, encapsulating them with a root property (in my case, I chose `state` as the name of the public property) is a must. Is there some way to set the root property that built-in validation uses?

I would like to use real-time validation (especially since this form can get so long), but I have to reference all of my validation like `state.field` (e.g.: `state.move_in_date`, `state.unit_id`, etc.). The side effect of this is that the default validation errors contain the word `state` in the message.

90% of the default messages are fine as-is, so I would only have to change a few (or set the attribute name to make the error message make sense). However, since they all contain the root property (`state`) I have to set every attribute's name. It seems like a lot of repeat work that shouldn't be necessary.

Laravel's validation doesn't have this problem, even when using `$this->validate()` in a controller, because you have to pass the data to validation. If I could pass the data myself, this wouldn't be a problem at all.

Stepping through the source, it looks like [`validateOnly()`](https://github.com/livewire/livewire/blob/3d0ce6c320bbfdfb6488f43033525bf607a5e613/src/ComponentConcerns/ValidatesInput.php#L200) (and `validate()`) call [`getDataForValidation()`](https://github.com/livewire/livewire/blob/3d0ce6c320bbfdfb6488f43033525bf607a5e613/src/ComponentConcerns/ValidatesInput.php#L355). This method (calls another method that) uses reflection to grab all the public properties and tries to use those for validation, comparing "before the first dot" to the property name. This is cool, but if I could short circuit this with a parameter to `validate*`, like in Laravel controllers, that would be awesome. This kind of behavior is already available for validation rules, so it isn't much of a leap to be able to provide my own array of data. If it is added as the last parameter, it could even be a non-breaking change. Thanks to PHP 8 I can use named parameters to put the data where I want it in an elegant way even.

I would be happy to make a PR for this change if requested.

**Comment by joshhanley** [comment, marked as answer]
@mikemand you could use custom attributes to change the name of the attributes as they are displayed to the user or you could just use a custom validator https://laravel-livewire.com/docs/2.x/input-validation#custom-validators

If they don't do what you're after, could you explain a bit more what you are hoping, and show an example of what you'd hope the validation code would look like?

Hope this helps!

**Reply by mikemand** [reply]
Hi Josh,

You're right, I totally missed the custom validators part of the docs. It would still be nice to have a shorthand, but this will work for my purposes. As a bonus, I can just copy them directly from the actions as they are. My app is built on Jetstream, so I was following their design patterns. However, it has become a big bloated mess. I could drop the interfaces and that would clear up the necessity of binding their implementations, but I would still have to go to a separate file for the logic of validating and saving forms. SRP is great for building small pieces of something (or a framework like Laravel), but a big monolithic app gets unwieldy with too much abstraction.

Here's how I was thinking of implementing a custom property for validation: https://github.com/mikemand/livewire/commit/b541ce5d4997e5393ba1acc2e3c6d5c2a6e06e07

I don't remember if the tests actually passed, but this will give you a rough idea. PHP 8 makes it even easier because with named parameters I can do this: `$this-validate(data: $this->myStateVariable);`, instead of having to pass `null, [], []` before the data.

**Reply by joshhanley** [reply]
@mikemand good to hear that will work 🙂 Ah yep I see how that would be helpful!

Yeah I'm not a fan of the structure of Jetstream personally, so just tend to use Breeze for everything instead, but you loose a lot of prebuilt features though 😕

**Reply by mikemand** [reply]
Yeah, that's why I went with Jetstream even though I don't like their structure. It saved me probably a month of work, so working around the annoying structure is worth it.

Do you want me to open a PR with my changes? There are some issues with the tests, but I don't actually understand why it can't find component 'foo'. Maybe I just missed something...

**Reply by joshhanley** [reply]
@mikemand yeah fair enough 🙂

Yeah go for it. Not sure what Caleb with think, he might just say use custom validators, but if you've already got the code there then it can't hurt to submit it!
